go-controller/pkg/ovn/loadbalancer/lb_cache.go

```diff
@@ -0,0 +1,376 @@
+	}
+
+	out := make([]CachedLB, 0, len(data.Data))
+
+	for _, row := range data.Data {
+		if len(row) != 5 {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: wrong number of columns (want 5) in row %#v", row)
+		}
+
+		res := CachedLB{
+			VIPs:     sets.String{},
+			Switches: sets.String{},
+			Routers:  sets.String{},
+		}
+		// parse the row
+
+		// name
+		if str, ok := row[0].(string); ok {
+			res.Name = str
+		} else {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: name: expected string, got %#v", row[0])
+		}
+
+		// uuid is a pair
+		if cell, ok := row[1].([]interface{}); ok {
+			if len(cell) != 2 {
+				return nil, fmt.Errorf("failed to parse find load_balancer response: uuid: expected pair, got %#v", cell)
+			} else {
+				if str, ok := cell[1].(string); ok {
+					res.UUID = str
+				} else {
+					return nil, fmt.Errorf("failed to parse find load_balancer response: uuid: expected string, got %#v", cell[1])
+				}
+			}
+		} else {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: uuid: expected slice, got %#v", row[1])
+		}
+
+		// protocol may be a string or an empty set
+		if str, ok := row[2].(string); ok {
+			res.Protocol = str
+		} else if _, ok := row[2].([]interface{}); ok { //empty set
+			// empty protocol, assume tcp
+			res.Protocol = "tcp"
+		} else {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: protocol: expected string, got %#v", row[2])
+		}
+
+		res.ExternalIDs, err = extractMap(row[3])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: external_ids: %w", err)
+		}
+
+		vips, err := extractMap(row[4])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse find load_balancer response: vips: %w", err)
+		}
+		for vip := range vips {
+			res.VIPs.Insert(vip)
+		}
+		out = append(out, res)
+	}
+
+	return out, nil
+}
+
+// findTableLBs returns a list of router name -> lb uuids
+func findTableLBs(tablename string) (map[string][]string, error) {
+	// this is CSV and not JSON because ovn-nbctl is inconsistent when a set has a single value :-/
+	rows, err := runNBCtlCSV([]string{"--data=bare", "--columns=name,load_balancer", "find", tablename})
+	if err != nil {
+		return nil, fmt.Errorf("failed to find existing LBs for %s: %w", tablename, err)
+	}
+
+	out := make(map[string][]string, len(rows))
+	for _, row := range rows {
+		if len(row) != 2 {
+			return nil, fmt.Errorf("invalid row returned when listing LBs for %s: %#v", tablename, row)
+		}
+		if row[0] == "" {
+			continue
+		}
+
+		lbs := strings.Split(row[1], " ")
+		if row[1] == "" {
+			lbs = []string{}
+		}
+		out[row[0]] = lbs
+	}
+
+	return out, nil
+}
+
+func TestOnlySetCache(cache *LBCache) {
+	globalCache = cache
+}
+
+// extractMap converts an untyped ovn json map in to a real map
+// it looks like this:
+// [ "map", [ ["k1", "v1"], ["k2", "v2"] ]]
+func extractMap(in interface{}) (map[string]string, error) {
```

Comment:
Should this go into the `util` package I feel like it my be useful in other places

Reply:
ugh noooo :-). All of this goes away when we have libovsdb. It's a silly hack to tide us over.